### ISSUE TITLE
Upgrade tests do not check pod instances and restart counts are identical for workload objects

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#78904]
**What happened**:

An upgrade of only the API server should not result in pod recreations or container restarts.

Today, this can accidentally be triggered by introducing new defaults to types within PodSpec (c.f. https://github.com/kubernetes/kubernetes/pull/69988, https://github.com/kubernetes/kubernetes/issues/69445, https://github.com/kubernetes/kubernetes/issues/78633).

The workload upgrade tests do not currently check if pod instances and container restart counts are identical.

**What you expected to happen**:

At the end of the upgrade test setup step, record the pod instances for the workload object in question, their uids, container restart counts, and associated node names and versions.

In the verification step post-upgrade, if the associated nodes still exist at the same versions, verify the same pod instances still exist with the same container restart counts.

/sig testing
/sig apps
/priority important-soon

**Comment by liggitt** [comment]
cc @smarterclayton @mml

**Comment by makoscafee** [comment]
@liggitt Hello! I'm bug triage shadow for the 1.16 release cycle and considering this issue is tagged for 1.16, but not updated for a long time, I'd like to check its status. The code freeze is starting on August 29th (about 1.5 weeks from now), which means that there should be a PR ready (and merged) until then.

Do we still target this issue to be fixed for 1.16?

**Comment by liggitt** [comment]
given there is no assigned owner, it does not appear so.

cc @kubernetes/sig-apps-misc for visibility on the need for stronger assertions in upgrade tests

**Comment by tnozicka** [comment]
I think such test would be valuable, I can take a look for next release
/assign

**Comment by smourapina** [comment]
@tnozicka @liggitt Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this issue still intended for 1.17?

**Comment by liggitt** [comment]
it is still important, and I would like to see it addressed in 1.17 if at all possible

**Comment by smourapina** [comment]
Friendly reminder that code freeze is rapidly approaching and will take place in November 14th. Is this issue still on track for 1.17? Otherwise we will need to move it to 1.18 milestone.

**Comment by dims** [comment]
@tnozicka still up for trying to do this in the next couple of days?

**Comment by smourapina** [comment]
@dimn @tnozicka Just a last reminder that we are pretty close to code freeze for 1.17 (tomorrow EOD PST).

**Comment by tnozicka** [comment]
I am sorry I won't make for this release, will try to add those in 1.18

**Comment by annajung** [comment]
Hello @tnozicka  👋 I'm one of the Bug Triage shadows for the 1.18 release. Since this issue is tagged for 1.18, I'd like to check its status. Is this issue intended for milestone 1.18?

Friendly reminder: Code freeze is scheduled for March 5th.

**Comment by annajung** [comment]
Hello @tnozicka, checking with you again. Is this issue intended for milestone 1.18?

**Comment by tnozicka** [comment]
Yes, I intend to work on those tests this week so we have it for 1.18

**Comment by annajung** [comment]
Hey @tnozicka checking in again as this issue is marked as `priority/important-soon`. Are you close to completing this for v1.18 release? Another reminder that code freeze is just around the corner! (March 5th).

**Comment by tnozicka** [comment]
@liggitt 

I agree with UIDs, revisions and dummy annotation. 

I am not yet convinced about the value of checking pod level properties on restart counts as upgrade is pretty much expected to restart the node or container runtimes. I am not sure the test can reliably determine which one of these changed.

The only reason I see to test it at workloads level is because we allowed a limited subset of PodSpec to be mutable and trigger container restarts.

**Comment by liggitt** [comment]
An upgrade of the API server is emphatically not expected to restart workloads. That's an important dimension to represent in our upgrade tests, and is what this issue is tracking.

**Comment by tnozicka** [comment]
the issue is tracking:

a) pod recreations
I think this is the majority of failures, at least those I've encountered with defaulting. Usually manifested as hotloops.

b) container restarts
Sounds more like a Pod level check. I could hook it there as well but I am unsure on  

> if the associated nodes still exist at the same versions

I haven't dig deep enough yet for what the upgrade actually does but real cluster upgrade procedure include upgrading runtimes, kubelet, rebooting nodes - which restarts pods or evicts them and place them to possibly different nodes (with new names). I am not sure what guarantees to expect from the upgrades in our suite so I didn't want to commit to testing something that might work only by accident. Wonder if we have any docs about upgrade tests guarantees.

**Comment by smarterclayton** [comment]
Great catch, this will make everything better.

**Comment by annajung** [comment]
Hey @tnozicka Just friendly reminder that code freeze for 1.18 release is in 2 days.

**Comment by annajung** [comment]
Hi there! Code freeze was yesterday for 1.18. I'll move this issue to v1.19 and remove it from bug triage tracking sheet. Thank you
/milestone v1.19

**Comment by liggitt** [comment]
cc @BenTheElder

**Comment by erismaster** [comment]
Seeing that this issue has had little activity in 1.20 milestone, I am going to remove the milestone. If this issue will be addressed in 1.21 please add it to the 1.21 milestone. 

/milestone clear

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/issues/134773 is another issue that should have been caught by upgrade tests.

1. Create one of every steadystate workload type (ReplicationController, ReplicaSet, Deployment, DaemonSet, StatefulSet)
2. Let the resulting pods get healthy and stabilize
3. Upgrade the control plane only (apiserver and kube-controller-manager)
4. Verify all workloads are still at the same revision / version and all pods are still at the same instance uids and restart counts
